### PR TITLE
*: fix cyclic replication issues with filter and ineligible tables

### LINKED ISSUE #678: If mark table not in allow list, create cyclic changefeed will fail
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do? If possible, provide a recipe for reproducing the error.

- create mark tables via cdc cli
- start a changefeed with normal tables in allowlist.

meet the error follows:

```
normal tables and mark tables are not paired
```

image

2. What did you expect to see?

create changefeed successfully

3. What did you see instead?

- Should the mark tables be included in table filter, if so, we need to update the document to emphasize it.
- Or if cyclic is enabled, the marketable should always be replicated regardless of the table filter config.

4. Versions of the cluster

    - Upstream TiDB cluster version (execute `SELECT tidb_version();` in a MySQL client):

        ```
        master
        ```

    - TiCDC version (execute `cdc version`):

        ```
        master
        ```

### LINKED ISSUE #679: Changefeed status does not forward normally in cyclic replication
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do? If possible, provide a recipe for reproducing the error.

- Setup two TiDB clusters in the A<->A deployment architecture, with multiple tables in upstream and downstream
- Start one cdc server in both TiDB cluster 
- create mark tables
- start changefeed in both cluster

2. What did you expect to see?

changefeed is replicated normally

3. What did you see instead?

- resolved ts and checkpoint ts is nor forwarded
- a lot of following logs

```
[2020/06/21 00:00:31.479 +08:00] [INFO] [etcd.go:110] ["update taskStatus"] [status="{\"tables\":{\"1029\":{\"start-ts\":417508562455756802,\"mark-table-id\":1646},\"1042\":{\"start-ts\":417508562455756802,\"mark-table-id\":1648},\"1107\"
:{\"start-ts\":417508562455756802,\"mark-table-id\":1656},\"1133\":{\"start-ts\":417508562455756802,\"mark-table-id\":1636},\"1193\":{\"start-ts\":417508562455756802,\"mark-table-id\":1638},\"1208\":{\"start-ts\":417508562455756802,\"mark
-table-id\":1652},\"1212\":{\"start-ts\":417508562455756802,\"mark-table-id\":1662},\"1217\":{\"start-ts\":417508562455756802,\"mark-table-id\":1664},\"1255\":{\"start-ts\":417508562455756802,\"mark-table-id\":1660},\"1625\":{\"start-ts\"
:417508562455756802,\"mark-table-id\":1650},\"49\":{\"start-ts\":417508562455756802,\"mark-table-id\":1668},\"53\":{\"start-ts\":417508562455756802,\"mark-table-id\":1654}},\"operation\":null,\"admin-job-type\":0}"]
[2020/06/21 00:00:31.481 +08:00] [INFO] [etcd.go:123] ["outdated table infos, ignore update taskStatus"]
[2020/06/21 00:00:31.848 +08:00] [INFO] [changefeed.go:374] ["dispatch table success"] [captureID=aa7362a6-e00d-46ec-940e-324f2aa94b75] [status="{\"tables\":{\"1029\":{\"start-ts\":417508562455756802,\"mark-table-id\":1646},\"1042\":{\"st
art-ts\":417508562455756802,\"mark-table-id\":1648},\"1107\":{\"start-ts\":417508562455756802,\"mark-table-id\":1656},\"1133\":{\"start-ts\":417508562455756802,\"mark-table-id\":1636},\"1193\":{\"start-ts\":417508562455756802,\"mark-table
-id\":1638},\"1208\":{\"start-ts\":417508562455756802,\"mark-table-id\":1652},\"1212\":{\"start-ts\":417508562455756802,\"mark-table-id\":1662},\"1217\":{\"start-ts\":417508562455756802,\"mark-table-id\":1664},\"1255\":{\"start-ts\":41750
8562455756802,\"mark-table-id\":1660},\"1625\":{\"start-ts\":417508562455756802,\"mark-table-id\":1650},\"49\":{\"start-ts\":417508562455756802,\"mark-table-id\":1668},\"53\":{\"start-ts\":417508562455756802,\"mark-table-id\":1654}},\"ope
ration\":{},\"admin-job-type\":0}"]
[2020/06/21 00:00:32.349 +08:00] [INFO] [changefeed.go:374] ["dispatch table success"] [captureID=aa7362a6-e00d-46ec-940e-324f2aa94b75] [status="{\"tables\":{\"1029\":{\"start-ts\":417508562455756802,\"mark-table-id\":1646},\"1042\":{\"st
art-ts\":417508562455756802,\"mark-table-id\":1648},\"1107\":{\"start-ts\":417508562455756802,\"mark-table-id\":1656},\"1133\":{\"start-ts\":417508562455756802,\"mark-table-id\":1636},\"1193\":{\"start-ts\":417508562455756802,\"mark-table
-id\":1638},\"1208\":{\"start-ts\":417508562455756802,\"mark-table-id\":1652},\"1212\":{\"start-ts\":417508562455756802,\"mark-table-id\":1662},\"1217\":{\"start-ts\":417508562455756802,\"mark-table-id\":1664},\"1255\":{\"start-ts\":41750
8562455756802,\"mark-table-id\":1660},\"1625\":{\"start-ts\":417508562455756802,\"mark-table-id\":1650},\"49\":{\"start-ts\":417508562455756802,\"mark-table-id\":1668},\"53\":{\"start-ts\":417508562455756802,\"mark-table-id\":1654}},\"ope
ration\":{},\"admin-job-type\":0}"]
[2020/06/21 00:00:32.848 +08:00] [INFO] [changefeed.go:374] ["dispatch table success"] [captureID=aa7362a6-e00d-46ec-940e-324f2aa94b75] [status="{\"tables\":{\"1029\":{\"start-ts\":417508562455756802,\"mark-table-id\":1646},\"1042\":{\"st
art-ts\":417508562455756802,\"mark-table-id\":1648},\"1107\":{\"start-ts\":417508562455756802,\"mark-table-id\":1656},\"1133\":{\"start-ts\":417508562455756802,\"mark-table-id\":1636},\"1193\":{\"start-ts\":417508562455756802,\"mark-table
-id\":1638},\"1208\":{\"start-ts\":417508562455756802,\"mark-table-id\":1652},\"1212\":{\"start-ts\":417508562455756802,\"mark-table-id\":1662},\"1217\":{\"start-ts\":417508562455756802,\"mark-table-id\":1664},\"1255\":{\"start-ts\":41750
8562455756802,\"mark-table-id\":1660},\"1625\":{\"start-ts\":417508562455756802,\"mark-table-id\":1650},\"49\":{\"start-ts\":417508562455756802,\"mark-table-id\":1668},\"53\":{\"start-ts\":417508562455756802,\"mark-table-id\":1654}},\"ope
ration\":{},\"admin-job-type\":0}"]
```

4. Versions of the cluster

    - Upstream TiDB cluster version (execute `SELECT tidb_version();` in a MySQL client):

        ```
        master
        ```

    - TiCDC version (execute `cdc version`):

        ```
        master
        ```

### PR DESCRIPTION
**Pull request by overvenus**
### What problem does this PR solve? 

This PR tries to:

- Fix #679 CDC stops replication due to ineligible tables are not ignored.
- Fix #678 If mark table not in allow list, create cyclic changefeed will fail

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

 - Has exported function/method change
 - Has exported variable/fields change

Side effects

 - Increased code complexity

Related changes

 - Need to cherry-pick to the release branch

### Release note

- Fix the issue that fail to create cyclic changefeed if mark table not in allow list.
- Fix the issue that cyclic changefeed does works if there are some ineligible tables.

### CODE REVIEW

File: cdc/owner.go [hunk 1]
```diff
@@ -240,6 +240,10 @@ func (o *Owner) newChangeFeed(
 			log.Warn("table not found for table ID", zap.Int64("tid", tid))
 			continue
 		}
+		if !tblInfo.IsEligible() {
+			log.Warn("skip ineligible table", zap.Int64("tid", tid), zap.Stringer("table", table))
+			continue
+		}
```
Comment (overvenus) [inline cdc/owner.go:246]: This change fixes #679 .

File: pkg/filter/filter.go [hunk 2]
```diff
@@ -74,6 +73,10 @@ func (f *Filter) ShouldIgnoreTable(db, tbl string) bool {
 	if IsSysSchema(db) {
 		return true
 	}
+	if f.isCyclicEnabled && mark.IsMarkTable(db, tbl) {
+		// Always replicate mark tables.
+		return false
+	}
```
Comment (overvenus) [inline pkg/filter/filter.go:79]: This change fixes #678 .

File: cdc/changefeed.go [hunk 3]
```diff
@@ -158,6 +158,11 @@ func (c *changeFeed) addTable(sid model.SchemaID, tid model.TableID, startTs mod
 		return
 	}
 
+	if !entry.WrapTableInfo(sid, table.Schema, tblInfo).IsEligible() {
+		log.Warn("skip ineligible table", zap.Int64("tid", tid), zap.Stringer("table", table))
+		return
+	}
```
Comment (overvenus) [inline cdc/changefeed.go:164]: This change fixes #679 .

File: cmd/util.go [hunk 4]
```diff
@@ -147,9 +147,10 @@ func verifyTables(ctx context.Context, cfg *config.ReplicaConfig, startTs uint64
 		if filter.ShouldIgnoreTable(tableName.Schema, tableName.Table) {
 			continue
 		}
-		fullTables = append(fullTables, tableName)
 		if !tableInfo.ExistTableUniqueColumn() {
```
Comment (leoppro) [inline cmd/util.go:150]:
```suggestion
		if !tableInfo. IsEligible() {
```

Comment (overvenus) [reply cmd/util.go:150]: `IsEligible()` filters view, is it ok?

File: cmd/util.go [hunk 5]
```diff
@@ -147,9 +147,10 @@ func verifyTables(ctx context.Context, cfg *config.ReplicaConfig, startTs uint64
 		if filter.ShouldIgnoreTable(tableName.Schema, tableName.Table) {
 			continue
 		}
-		fullTables = append(fullTables, tableName)
-		if !tableInfo.ExistTableUniqueColumn() {
+		if !tableInfo. IsEligible() {
```
Comment (overvenus) [inline cmd/util.go:150]:
```suggestion
		if !tableInfo.IsEligible() {
```

**Pull request merged**